components/calendar/src/julian.rs

```diff
@@ -0,0 +1,317 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! This module contains types and implementations for the Julian calendar
+
+use crate::iso::{Iso, IsoDateInner, IsoDay, IsoMonth, IsoYear};
+use crate::{types, Calendar, Date, DateDuration, DateDurationUnit, DateTime, DateTimeError};
+use core::convert::TryInto;
+
+#[derive(Copy, Clone, Debug, Default)]
+// The Julian calendar
+pub struct Julian;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+// The inner date type used for representing Date<Georgian>
+pub struct JulianDateInner(IsoDateInner);
+
+impl Calendar for Julian {
+    type DateInner = JulianDateInner;
+    fn date_from_iso(&self, iso: Date<Iso>) -> JulianDateInner {
+        let added_days = Self::calculate_day_difference_between_calendars(*iso.inner());
+        Self::offset_date(
+            &mut JulianDateInner(*iso.inner()),
+            DateDuration::new(0, 0, 0, -added_days),
+        )
+    }
+
+    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
+        let added_days = Self::calculate_day_difference_between_calendars(date.0);
+        let mut date = Self::offset_date(
+            &mut JulianDateInner(date.0),
+            DateDuration::new(0, 0, 0, added_days),
+        );
+
+        // Edge case when the year is divisible by 100, and not by 400
+        if date.0.year.0 % 100 == 0
+            && date.0.year.0 % 400 != 0
+            && u8::from(date.0.month) == 2
+            && u8::from(date.0.day) == 29
+        {
+            date.0.month = 3.try_into().unwrap();
+            date.0.day = 1.try_into().unwrap();
+        }
+
+        Date::from_raw(date.0, Iso)
+    }
+
+    fn months_in_year(&self, date: &Self::DateInner) -> u8 {
+        Iso.months_in_year(&date.0)
+    }
+
+    fn days_in_year(&self, date: &Self::DateInner) -> u32 {
+        if Self::is_leap_year(date.0.year) {
+            366
+        } else {
+            365
+        }
+    }
+
+    fn days_in_month(&self, date: &Self::DateInner) -> u8 {
+        Self::days_in_month(date.0.year, date.0.month)
+    }
+
+    fn day_of_week(&self, date: &Self::DateInner) -> types::IsoWeekday {
+        Iso.day_of_week(Julian.date_to_iso(date).inner())
+    }
+
+    fn offset_date(&self, date: &mut Self::DateInner, offset: DateDuration<Self>) {
+        Self::offset_date(date, offset);
+    }
+
+    #[allow(clippy::field_reassign_with_default)]
+    fn until(
+        &self,
+        date1: &Self::DateInner,
+        date2: &Self::DateInner,
+        largest_unit: DateDurationUnit,
+        smallest_unit: DateDurationUnit,
+    ) -> DateDuration<Self> {
+        Iso.until(&date1.0, &date2.0, largest_unit, smallest_unit)
```

Comment:
issue: we can't actually use this because the durations mean different things

we should do the same code copying as offset_date here